### DISCUSSION TITLE
Draw polygons using libvips

### DISCUSSION THREAD

**Discussion by ido-nuc** [libvips/libvips#3560] [Q&A] [answered]
Hi!

I am currently working on a pathology-related project where I am dealing with (very) large-sized images. I chose `libvips` (specifically the `pyvips` binding) due its capability to handle large images efficiently. My pipeline includes rendering shapely geometries (i.e., points and polygons) & blending multiple rendered images together.

However, after searching through the `libvips` documentation, I noticed that there doesn't seem to be a `draw_polygon` operation. This leads me to two questions:

1. Why `libvips` does not have a `draw_polygon` operation? If it does exist and I missed it, please let me know.
2. Given the current set of operations, what would you suggest as the best way to implement the pipeline I discussed above? Any advice on how to approach this and any alternative methods for rendering shapely geometries and blending images together would be greatly appreciated.

Thank you for your assistance!

Best,
Ido

**Comment by jcupitt** [comment, marked as answer]
Hi @ido-nuc,

The simplest solution is just to draw them with SVG:

```python
import pyvips

x = pyvips.Image.svgload_buffer(b"""
    <svg viewBox="0 0 200 200">
      <circle r="100" cx="100" cy="100" fill="#900"/>
    </svg>
""")

x.write_to_file("x.png")
```

To make:

![circle](https://github.com/libvips/libvips/assets/580843/936be870-2bd9-46eb-8e10-5615cda66c5c)

The SVG loader is progressive, so it can draw SVGs of any size (up to 32k x 32k pixels) using only a modest amount of memory. You can make masks by just taking the alpha, you can combine masks with the boolean operators, and you can render stacks of RGBA images with `composite`.

The 32k pixel limit is unfortunate :( It comes from cairo, sadly, and isn't really fixable. We are considering switching to another SVG renderer, but no one's done the work yet.

Don't use the `draw_circle` etc. operators, They do not work well with large images.

Another option is to use `xyz`. This makes an n-dimensional index image where each pixel contains its own coordinates. You can then use the usual arithmetic, relational and boolean operators to draw any shape. It's surprisingly quick, and scales well to images to any size. Coming up with the right maths to draw the shape you want is annoying, however.

```python
x = pyvips.Image.xyz(501, 501) - [250, 250]
x = (x[0] ** 2 + x[1] ** 2) ** 0.5 < 200
```

Will draw a circle with a radius of 200 in the centre of a 500 x 500 pixel image, for example.

**Reply by lejsekt** [reply]
In case somebody needs a similar operation in C#:

```csharp
using var x = Image.Xyz(501, 501) - new[] { 250, 250 };
using var distance = x[0] * x[0] + x[1] * x[1];
using var pow = distance.Pow(0.5);
using var result = pow <= 200;
```
